agb/src/display/tiled/map.rs

```diff
@@ -77,86 +224,165 @@ impl RegularMap {
-        for tile in self.tiles.iter_mut() {
-            if *tile != Tile::default() {
-                vram.remove_tile(tile.tile_index());
-            }
-
-            *tile = Tile::default();
-        }
+    fn x_register(&self) -> MemoryMapped<i16> {
+        unsafe { MemoryMapped::new(0x0400_0010 + 4 * self.background_id as usize) }
     }
 
-    pub fn show(&mut self) {
-        let mode = DISPLAY_CONTROL.get();
-        let new_mode = mode | (1 << (self.background_id + 0x08));
-        DISPLAY_CONTROL.set(new_mode);
+    fn y_register(&self) -> MemoryMapped<i16> {
+        unsafe { MemoryMapped::new(0x0400_0012 + 4 * self.background_id as usize) }
     }
+}
 
-    pub fn hide(&mut self) {
-        let mode = DISPLAY_CONTROL.get();
-        let new_mode = mode & !(1 << (self.background_id + 0x08));
-        DISPLAY_CONTROL.set(new_mode);
-    }
+pub struct AffineMap {
+    background_id: u8,
+    screenblock: u8,
+    priority: Priority,
+    size: AffineBackgroundSize,
 
-    pub fn commit(&mut self, vram: &mut VRamManager) {
-        let new_bg_control_value = (self.priority as u16)
-            | (u16::from(self.screenblock) << 8)
-            | (self.size.size_flag() << 14);
+    scroll: Vector2D<i16>,
 
-        self.bg_control_register().set(new_bg_control_value);
-        self.bg_h_offset().set(self.x_scroll);
-        self.bg_v_offset().set(self.y_scroll);
+    transform_origin: Vector2D<Num<i32, 8>>,
+    transform: BgAffineSetData,
 
-        let screenblock_memory = self.screenblock_memory();
+    tiles: Vec<u8>,
+    tiles_dirty: bool,
+}
 
-        if self.tiles_dirty {
-            unsafe {
-                dma_copy16(
-                    self.tiles.as_ptr() as *const u16,
-                    screenblock_memory,
-                    self.size.num_tiles(),
-                );
-            }
-        }
+impl TiledMapTypes for AffineMap {
+    type Size = AffineBackgroundSize;
+}
 
-        vram.gc();
+impl TiledMapPrivate for AffineMap {
+    type TileType = u8;
+    type AffineMatrix = AffineMatrixAttributes;
 
-        self.tiles_dirty = false;
+    fn tiles_mut(&mut self) -> &mut [Self::TileType] {
+        &mut self.tiles
     }
-
-    pub fn set_scroll_pos(&mut self, pos: Vector2D<u16>) {
-        self.x_scroll = pos.x;
-        self.y_scroll = pos.y;
+    fn tiles_dirty(&mut self) -> &mut bool {
+        &mut self.tiles_dirty
     }
-
-    #[must_use]
-    pub fn scroll_pos(&self) -> Vector2D<u16> {
-        (self.x_scroll, self.y_scroll).into()
+    fn background_id(&self) -> usize {
+        self.background_id as usize
     }
-
-    pub(crate) fn size(&self) -> RegularBackgroundSize {
+    fn screenblock(&self) -> usize {
+        self.screenblock as usize
+    }
+    fn priority(&self) -> Priority {
+        self.priority
+    }
+    fn map_size(&self) -> Self::Size {
         self.size
     }
+    fn update_bg_registers(&self) {
+        let register_pos = self.transform.position + self.transform_origin;
```

Comment:
```suggestion
        let register_pos = self.transform.position;
```

The calculation of the matrix already takes this into account.